src/models/transaction.rs

```diff
@@ -1,197 +1,161 @@
     }
 
-    test_rpc_to_primitive_conversion!(
-        test_legacy_transaction_conversion_to_primitive,
-        legacy_rpc_transaction,
-        gas_price,
-        false
-    );
-
-    test_rpc_to_ec_recovered_conversion!(
-        test_legacy_transaction_conversion_to_ec_recovered,
-        legacy_rpc_transaction,
-        gas_price,
-        false
-    );
-
-    test_rpc_to_primitive_conversion!(
-        test_eip2930_transaction_conversion_to_primitive,
-        eip2930_rpc_transaction,
-        gas_price,
-        true
-    );
-    test_rpc_to_ec_recovered_conversion!(
-        test_eip2930_transaction_conversion_to_ec_recovered,
-        eip2930_rpc_transaction,
-        gas_price,
-        true
-    );
-
-    test_rpc_to_primitive_conversion!(
-        test_eip1559_transaction_conversion_to_primitive,
-        eip1559_rpc_transaction,
-        max_fee_per_gas,
-        true
-    );
-
-    test_rpc_to_ec_recovered_conversion!(
-        test_eip1559_transaction_conversion_to_ec_recovered,
-        eip1559_rpc_transaction,
-        max_fee_per_gas,
-        true
-    );
+    let mut calldata = Vec::with_capacity(capacity);
+
+    // assert that the selector < FieldElement::MAX - retries
+    assert!(*ETH_SEND_TRANSACTION < FieldElement::MAX - retries.into());
+    let selector = *ETH_SEND_TRANSACTION + retries.into();
+
+    // Retries are used to alter the transaction hash in order to avoid the
+    // DuplicateTx error from the Starknet gateway, encountered whenever
+    // a transaction with the same hash is sent multiple times.
+    // We add the retries to the selector in the calldata, since the selector
+    // is not used in by the EOA contract during the transaction execution.
+    calldata.append(&mut vec![
+        FieldElement::ONE,        // call array length
+        *KAKAROT_ADDRESS,         // contract address
+        selector,                 // selector + retries
+        FieldElement::ZERO,       // data offset
+        signed_data.len().into(), // data length
+        signed_data.len().into(), // calldata length
+    ]);
+    calldata.append(&mut signed_data);
+
+    Ok(calldata)
+}
+
+pub(crate) async fn transaction_retries(
+    transaction_signed: &TransactionSigned,
+    db: &Database,
+) -> Result<u8, EthApiError> {
+    let hash = transaction_signed.hash;
+    // Get the pending transaction from the database and extract the retries
+    let filter = into_filter("tx.hash", &hash, HASH_HEX_STRING_LEN);
+    Ok(db
+        .get_one::<StoredPendingTransaction>(filter.clone(), None)
+        .await?
+        .map(|tx| tx.retries + 1)
+        .inspect(|retries| tracing::info!("Retrying {} with {} retries", hash, retries))
+        .or_else(|| {
+            tracing::info!("New transaction {} in pending pool", hash);
+            None
+        })
+        .unwrap_or_default())
+}
+
+pub(crate) async fn update_pending_transaction_in_database(
```

Comment:
What do you think about migrating this function to the `Database` structure to minimize functions which are not inside objects? 

So that we can call: `database.update_pending_transaction(...`

Reply:
I thought about this do but wasn't a great fan of the solution, because for now `Database` is a wrapper around the mongo db type which provides some light utilities for us, but isn't "aware" of Ethereum types. Will think of something nicer maybe

Reply:
Yes I understand this point of view too. This is already a lot cleaner like this, so if there is no easy solution to wrap the functions inside objects, let us keep like that.

Reply:
I'll think of something, would be nice to have this removed indeed